signer/src/block_observer.rs

```diff
@@ -400,15 +397,15 @@ mod tests {
     }
 
     impl StacksInteract for TestHarness {
-        async fn get_blocks_by_bitcoin_block(
+        async fn get_last_tenure_blocks(
```

Comment:
While I can see why it's easier to get the information from the last tenure and not having it connected to a bitcoin block, I'm also afraid that will open up the possibility of a whole class of inconsistencies. What if the stacks node observes a different canonical fork than the block notifier? What if we're backfilling old bitcoin blocks and not working on the "last" bitcoin block hash?

The idea of connecting the two is to ensure that all signers who observe the same bitcoin block has a consistent view of both blockchains.

Reply:
All great points. I have some ideas.

> What if the stacks node observes a different canonical fork than the block notifier?

This is a semi-tough problem that exists in both setups. If our block notifier disagrees with our stacks node on the canonical fork then we have to wait for the issue to resolve itself, which will happen eventually. In each case, the stacks-node will have a different idea on what has been confirmed and what hasn't.

> What if we're backfilling old bitcoin blocks and not working on the "last" bitcoin block hash?

Because Stacks will fork with bitcoin, we start by getting the blocks for the last tenure make multiple `GET /v3/tenures/<block-id>` requests (`StacksClient::get_tenure` requests) until we get blocks that match what we have in the database.

---

The real reason why I moved away from the bitcoin block based API is because I couldn't get things to work (the Hiro API is not up to date with Nakamoto changes). But I think this approach works just as well.

Reply:
I'd agree with Mårten on this one if this weren't for testing, but since this is just a test I think it makes the most sense to get the latest tenure blocks and then find a way to verify that they correspond to the correct Bitcoin block before continuing with the test function. It'd really suck for this inconsistency to make us think something else in the code is wrong.

Maybe keep the `bitcoin_block_hash` variable in the function call and if it's feasible add an assert at the end that forces the the tenure to be pegged to that bitcoin block hash.

Reply:
While this implementation is for testing, the interface is the one used in the block observer - which is why this has design implications. Daniel and I had a chat about this and the problems with the Hiro API and Nakamoto blocks, we decided that the block observer should consume a stream of stacks block headers, independent of the stream of bitcoin block headers it currently consumes from the bitcoin block notifier - and it will use both of these streams to populate the signer state.